dashboard/app/controllers/pd/workshop_certificate_controller.rb

```diff
@@ -6,6 +6,8 @@ class Pd::WorkshopCertificateController < ApplicationController
     find_by: :code, id_param: :enrollment_code
 
   def generate_certificate
+    return render_404 unless @enrollment&.attendances&.any?
```

Comment:
I think the first safe-navigation operator isn't needed here because the `load_resource` line above that populates `@enrollment` should cause a 404 if it's not found before we reach this code; and the second safe-navigation operator shouldn't be necessary because if you have an `Enrollment` with no attendances, its `attendances` association will give an empty array, not `nil`. But use tests to check these cases before taking my word for it!

```suggestion
    return render_404 unless @enrollment.attendances.any?
```

Reply:
Makes sense -- we do have a [test](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/controllers/pd/workshop_certificate_controller_test.rb#L27) that an error is raised if a nonsense enrollment code is provided, so I think your first assumption is sound. I tested manually in console that an enrollment with no attendances returns false from `enrollment.attendances.any?`, not an error. Also tested the `generate_workshop_certificate` page with this code change and an enrollment with no attendance got the 404 I'd expect.